Fetch new games

janakerman opened the pull request:
At the moment the latest games are just stored in a hashmap in the EntityService. We'll want to persist this at some point in the future when we start persisting our data.

janakerman committed:
Games are now only fetched if they haven't been seen before.

janakerman committed:
Extracts fetching of a dataset to dedicated object

The LichessResultSet encapsulates the logic for fetching an
entire dataset spread over multiple pages. It's generic and
works for both Players and Games.

bjedrzejewski commented on src/main/java/com/chessmates/service/EntityServiceImpl.groovy:
```diff
@@ -31,60 +33,80 @@ class EntityServiceImpl implements EntityService {
+                def games = resultSet.get()
+
+                if (games.size()) {
+                    latestGameMap.putIfAbsent(playerCombination, games.first())
```
Jan says it's a mistake

bjedrzejewski commented on src/main/java/com/chessmates/utility/LichessResultSet.groovy:
```diff
@@ -0,0 +1,75 @@
+        def items = []
+
+        // Start at page one and keep requesting pages until there are no more pages.
+        def previousPage = getPage.apply(1)
```
This could be called currentPage, as this is what is about to get requested

bjedrzejewski commented on src/main/java/com/chessmates/utility/LichessResultSet.groovy:
```diff
@@ -0,0 +1,75 @@
+        items.addAll(getItemsFromPage(previousPage, shouldStop)['items'])
+
+        // Loop around all following pages until the stop condition is fulfilled or we hit the end of the results.
+        while(previousPage?.nextPage != null) {
```
You can drop off !=null since we are in the Groovy world

bjedrzejewski replied:
I don't see a place where you re-assign the previousPage variable. Also, are you sure that you need Elvis there?

janakerman replied:
For the same reason as below I'm going to leave it an explicit null check. This could conceivably work with negative page numbers, and in that case 0 would be a valid option.

bjedrzejewski commented on src/main/java/com/chessmates/utility/LichessResultSet.groovy:
```diff
@@ -0,0 +1,75 @@
+        def pageItems = []
+        def iterator = page.results.iterator()
+        T item
+        while ((item = iterator[0])) {
```
I think there are extra brackets there

janakerman replied:
I've changed this back to checking for null. Since the class is generic, it could very well be working with Integer objects. In that case, the value 0 would be falsy, even though its a completely legitimate value.

`while ((item = iterator[0]) != null) {` << that's what I'm putting now

bjedrzejewski reviewed (changes requested):
We need to test and ensure that the logic problems are addressed

janakerman committed:
Adds unit tests for LichessResultsSet

janakerman committed:
Addresses mistake in review comments

We always want to put this - otherwise we'll never update the reference to the
newest game.

janakerman commented:
High level tests on EntityServiceImpl are incoming, but I think we can merge this now.

bjedrzejewski reviewed (approved):
👍 🥇

janakerman committed:
Extracts function from LichessResultSet into own file

It wasn't worth the hassle!

janakerman committed:
Finalize various propertise

janakerman commented:
Don't like the giant constructors on `LichessResultPage`, turns out you can't use .with {} on private properties, and since we're not using Groovy 2.5 it doesn't have all the nice @Builder annotations. I'm sure there's 101 builder/factory patterns to clean this up but not worth the time at the moment imo!

janakerman commented:
Also not sure about all the casting in LichessApiImpl - opinions there would be good. Going to merge anyway as I still think it's an improvement on what was already there.

janakerman commented:
Next work going forwards is testing all of the above properly and splitting out the 'get games' functionality from the 'scrape data from lichess' functionality as these will differ when we get the DB connected.
